enhancements/installer/compact-plus-worker.md

```diff
@@ -0,0 +1,194 @@
+---
+
+# Compact + Worker deployment as day0 operation
+
+## Summary
+
+This enhancement proposes to allow for deploying a 4-node cluster as a day 0
+operation, where the infrastructure workload can be deployed with HA enabled.
+The 4-nodes cluster would be composed by 3 schedulable master nodes and a
+single worker, which provide users with a highly available cluster.
+
+## Motivation
+
+There are environments where only 4 nodes are available where 4-nodes clusters
+are being deployed. We could serve OpenShift cluster administrators of these
+environments better by allowing them to deploy a 4-nodes cluster as a day0
+operation and make sure their infrastructure workload is highly available.
+
+Today, it is possible to deploy a 4-nodes cluster. The resulting cluster will
+be functional but the infrastructure workload will be run as a SingleReplica
+topology, which won't allow the various services to run as HA.
+
+To achieve the desired topology described in this enhancement,OpenShift cluster
+administrators can only deploy a 4-nodes clusters by first deploying a compact
+cluster (3 schedulable masters) and then adding a host as a day 2 operation.
+Another option is for them to <missing text here> This provides the same results as aimed in this
+enhancement but with a poor and slower user experience.
+
+Today, because of the current limitations in the installer, it takes
+administrators significantly longer to achieve the desired topology, when only
+4 nodes are available. Furthermore, by requiring more steps, we are making the
+process more prone to user errors.
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+
+- Enable the installer to deploy a 4-nodes cluster composed by 3 schedulable
+  masters and one worker.
+
+### Non-Goals
+
+- Define new infrastructure topologies
+
+- Define new architectures
+
+- Provide official support for the 4-nodes cluster deployments
+
+## Proposal
+
+- Set the `infrastructureTopology` to have the same value as the
+  `controlPlaneTopology` when the number of workers is 1, instead of setting its
+  value to `SingleReplica` as it's done today.
+
+- Set masters as schedulable when workers are less or equal to 1. Today,
+  masters are set to schedulable only when there are no workers available.
+
+### User Stories
+
+- As an OpenShift cluster administrator, I want to be able to install a 4-nodes
+  cluster as a single installation step so that my cluster will be able to
+  handle workloads that require replication.
+
+
+### API Extensions
+
+- Sets the `infrastructureTopology` field to the value of `controlPlaneTopology` in the
+  `config.openshift.io/v1/infrastructure` CR called `cluster`, which is created
+  during the install process, when the number of workers is 1.
+
+- Sets the `mastersSchedulable` field to true in the
+  `config.openshift.io/v1/scheduler` CR called `cluster`, which is created
+  during the install process, when the number of workers is 1.
+
```

Comment:
I pointed out the relationship to #1041 in an earlier comment. There the proposal is to add a new API field to control how ingress is deployed because even in an environment where we could run 2 controllers (both nodes are schedulable), we want to keep ingress coming through the control plane node.

In this proposal, instead of saying the infrastructureTopology is the same as the controlPlaneTopology, we could say that rather than counting only workers the infrastructureTopology setting should be based on the _total number of nodes on which infrastructure pods could be scheduled_ (including the control plane, when those nodes are schedulable).

In a SNO configuration, the field would be set to singleReplica, as it is today.

In a SNO+1 configuration, after this change the field would be set to HA. If we accept the API proposed in #1041, we would keep ingress off of the worker node, but other infrastructure pods could run there. Is that desirable?

In a compact cluster, both fields would be set to HA. I believe that's consistent with the current behavior, right?

In a 3+1 cluster, both fields would be set to HA. That's the goal of this enhancement.

In an HA cluster, both fields would be set to HA. That's consistent with the current behavior.

Reply:
> In this proposal, instead of saying the infrastructureTopology is the same as the controlPlaneTopology, we could say that rather than counting only workers the infrastructureTopology setting should be based on the total number of nodes on which infrastructure pods could be scheduled (including the control plane, when those nodes are schedulable).

This is the gist of this proposal and what I think the final change would look like.

Admittedly, this enhancement could be rephrased in terms of the changes desired on the way the InfrastructureTopology is chosen. I decided to define the enhancement around the 4-nodes "concept" because of the changes proposed to the way we automatically flag masters as schedulable.

I think both changes are important. Just doing the change to how the InfrastructureTopology is chosen would still allow people to create a 4-nodes cluster that is unhealthy, which would result in a bad user experience.

Reply:
@dhellmann @staebler @hardys thanks for your comments so far. 

I have been looking more into this and what options are available for the user.

I wonder if it would be better for the discussion to narrow things a bit more and focus on the part that would, at least, unblock the functionality. There are 2 items that I think would help us move forward. These items are not mutually exclusive, but I think they can be worked on independently:

- Allow for setting the `InfrastructureTopology` through manifests:
    - I wrote down my findings in my last comment on [this thread](https://github.com/openshift/enhancements/pull/1057/files?diff=unified&w=1#r824987520). Unless I am missing something, there is currently no way for the `InfrastructureTopology` to be set through manifests
- Count schedulable masters as workers when selecting the InfrastructureTopology


The above leaves out, from this enhancement, the switch of the master nodes to schedulable when there is one worker. This will still need to be an explicit action (override on the manifests) done by the user.

Since setting masters as schedulable requires modifying a manifest, it may make more sense allow for setting the InfrastructureTopology through manifests too. This, however, means API changes to the Infrastructure CR will be required. Again, this is all assuming my findings are correct.

In my opinion, I think overriding manifests is not the greatest use experience. I understand the value as it helps with providing a more explicit support matrix. On the other hand, making assumptions like 1 worker == SingleReplica (especially when they cannot be changed after the fact) is also not a good experience.

Reply:
I took more time to do an e2e test with assisted installer and manifests override. I managed to get a 4-node cluster deployed but there are multiple limitations with the current approach. 

1. The current approach requires applying custom manifests
2. openshift-installer applies manifests in order rather than merging them. This means that overriding the Infrastructure CR requires passing a fully populated manifest. 
3. The InfrastructureTopology is a `Status` field, which is commonly read as an output that came from some internal operations. Overriding it just doesn't feel right.

The way custom manifests is implemented in Assisted Service is that the custom manifest is pushed to the API, then it's made part of the bootstrap ignition. In my tests I created a file called `55-infrastructure-override.yaml` that was added to the `manifests` dir that the `openshift-installer` read from. This manifest overwrote the original one generated by the installer.

I got it to work in my test but I don't think this is a good approach. There are fields in the Infrastructure CR that are computed and they are dynamic (i.e infrastructureName). To override the Infrastructure CR, we would have to develop an infrastructure name that will differ from the one generated by the installer. Here's an example of what it looks like:

![image](https://user-images.githubusercontent.com/13816/161787817-ed31845b-a88f-4ede-9bf6-f586513c5e14.png)

In the screenshot above, the infrastructure CR has the name I picked but the machines were created with the name computed by the installer.

Looking ahead (and taking a step back):

I think as a first approach we could start by exposing the `InfrastructureTopology` somehow through the deployment process. Either by putting it in the install-config or by exposing it in the Infrastructure CRD spec. We can leave any conversation about setting masters as schedulable and counting schedulable masters as workers for future enhancements.

Would this be something reasonable? @staebler @dhellmann 

The impact of this work would be smaller and it would allow automated flows to customize deployments without having to consider this architecture supported or having to expand the test matrix. 

For completeness, here's a screenshot showing the deployment tasks, the nodes, and operators:

![image](https://user-images.githubusercontent.com/13816/161789022-48758c5b-c2ba-43cc-9200-b08a12e2eea8.png)

Reply:
> 1. The current approach requires applying custom manifests

This is an accepted and valid way of interacting with the OpenShift installer.

> 2. openshift-installer applies manifests in order rather than merging them. This means that overriding the Infrastructure CR requires passing a fully populated manifest.

The OpenShift installer supports editing the manifests that the OpenShift Installer generates. The OpenShift Installer does not require, or even condone, the user to provide a fully populated manifest.

> 3. The InfrastructureTopology is a Status field, which is commonly read as an output that came from some internal operations. Overriding it just doesn't feel right.

The fact that it is a Status field is a consequence of using the difficulty of updating Status fields as a placeholder for immutability. In this case, there is no restriction against or recommendation to not edit the Status fields.

> The way custom manifests is implemented in Assisted Service is that the custom manifest is pushed to the API, then it's made part of the bootstrap ignition.

Can the Assisted Service be modified to support editing manifests in addition to the current behavior of overwriting manifests?

Reply:
> > 1. The current approach requires applying custom manifests
> 
> This is an accepted and valid way of interacting with the OpenShift installer.

I agree, manifests are the preferred way of doing "unusual" things like this.

> > 2. openshift-installer applies manifests in order rather than merging them. This means that overriding the Infrastructure CR requires passing a fully populated manifest.
> 
> The OpenShift installer supports editing the manifests that the OpenShift Installer generates. The OpenShift Installer does not require, or even condone, the user to provide a fully populated manifest.
> 
> > 3. The InfrastructureTopology is a Status field, which is commonly read as an output that came from some internal operations. Overriding it just doesn't feel right.
> 
> The fact that it is a Status field is a consequence of using the difficulty of updating Status fields as a placeholder for immutability. In this case, there is no restriction against or recommendation to not edit the Status fields.

I'm not sure I'd go quite that far. We could say that for this use case it's OK to change the value in the manifest, but I'm not sure we want to recommend that users do that themselves as a regular practice. I know we can't stop them from changing it, but that's different from recommending it.

> > (dhellmann) In this proposal, instead of saying the infrastructureTopology is the same as the controlPlaneTopology, we could say that rather than counting only workers the infrastructureTopology setting should be based on the total number of nodes on which infrastructure pods could be scheduled (including the control plane, when those nodes are schedulable).
> 
> (flaper87) This is the gist of this proposal and what I think the final change would look like.
> 
> Admittedly, this enhancement could be rephrased in terms of the changes desired on the way the InfrastructureTopology is chosen. I decided to define the enhancement around the 4-nodes "concept" because of the changes proposed to the way we automatically flag masters as schedulable.
> 
> I think both changes are important. Just doing the change to how the InfrastructureTopology is chosen would still allow people to create a 4-nodes cluster that is unhealthy, which would result in a bad user experience.

I don't think you fully answered my earlier question. It seems to me that the problem isn't the actual number of nodes being deployed, but is rather that the way we compute the value for the infrastructureTopology is incorrect because it doesn't count all of the nodes that it should.

I think that change, combined with ZTP explicitly setting the control plane nodes to be schedulable, should be sufficient for what you're trying to do in this 4 node case and that it would work correctly for other size clusters, too.

Would there be any negative side-effects of using that approach?

Reply:
> Can the Assisted Service be modified to support editing manifests in addition to the current behavior of overwriting manifests?
Nice. Never thought in this direction (and honestly, no one ever needed this functionality so far). But worth exploring regardless (unrelated to this enhancement of course). BTW, I assume hive does not support editing neither, right?

Reply:
> > Can the Assisted Service be modified to support editing manifests in addition to the current behavior of overwriting manifests?
> 
> Nice. Never thought in this direction (and honestly, no one ever needed this functionality so far). But worth exploring regardless (unrelated to this enhancement of course). BTW, I assume hive does not support editing neither, right?

You are correct that Hive does not current support editing either. They have a use case for it to support Managed ODF, where they would want to edit the worker machineset manifests to add a security group.

Reply:
> > > (flapper87) 5\. The InfrastructureTopology is a Status field, which is commonly read as an output that came from some internal operations. Overriding it just doesn't feel right.
> > 
> > 
> > (staebler) The fact that it is a Status field is a consequence of using the difficulty of updating Status fields as a placeholder for immutability. In this case, there is no restriction against or recommendation to not edit the Status fields.
> 
> (dhellman) I'm not sure I'd go quite that far. We could say that for this use case it's OK to change the value in the manifest, but I'm not sure we want to recommend that users do that themselves as a regular practice. I know we can't stop them from changing it, but that's different from recommending it.

Note that the user is not mutating the Status of a resource. The user is editing the Status of a manifest representing a resource to be created. The Status is not representative of the status of a resource, which is something that I agree we would to want to suggest that a user should edit. By the nature of installer manifests, none of the Status fields in the manifests are representative of the status of a resource. I see no meaningful difference between the user editing a Spec field or a Status field of the manifest for the Infrastructure resource.

Reply:
> > (flaper87) 2. openshift-installer applies manifests in order rather than merging them. This means that overriding the Infrastructure CR requires passing a fully populated manifest.
> 
> (staeblerer) The OpenShift installer supports editing the manifests that the OpenShift Installer generates. The OpenShift Installer does not require, or even condone, the user to provide a fully populated manifest.

Can you point me to an example of how I can pass a partially-populated manifest to the installer? 

I tried passing an Infrastructure CR with just the `infrastructureTopology` and it ended up overriding the entire CR, leaving all the other fields in the status empty.
 
> > (flaper87) The way custom manifests is implemented in Assisted Service is that the custom manifest is pushed to the API, then it's made part of the bootstrap ignition.
> 
> (staeblerer) Can the Assisted Service be modified to support editing manifests in addition to the current behavior of overwriting manifests?

I thought of this but I didn't want to "just do it on our side" but rather bring the discussion here so that we could have a chance to think whether it would make sense to expose it through the installer or not. I hear that it's ok to implement our own patching and logic, which would help us work around this limitation.

> (dhellmann) I don't think you fully answered my earlier question. It seems that the problem isn't the actual number of nodes being deployed, but rather that the way we compute the value for the infrastructureTopology is incorrect because it doesn't count all of the nodes that it should.
> 
> I think that change, combined with ZTP explicitly setting the control plane nodes to be schedulable, should be sufficient for what you're trying to do in this 4 node case and that it would work correctly for other size clusters, too.
>
> Would there be any negative side-effects of using that approach?

It would be sufficient as it would pick the right InfrastructureTopology. This would bring us back to the points @hardys brought up with regards to what is tested and what isn't. Right now it seems that only compact and full cluster deployments are being tested. 

This is fundamentally what I am trying to achieve in this proposal, to have the installer take into account schedulable masters and workers when picking the topology. Admittedly my wording was not the best.

If this is fine and an acceptable solution, I would be happy to reword the enhancement so that we can focus on that, without exposing the `InfrastructureTopology`. The reason I proposed the latter in my previous comment is that I wanted the installer to make fewer assumptions about what the environment looks like and what the customer wants. especially for fields that are immutable for the rest of the cluster lifecycle.